Bottom: 56e1cb827a694ea18772a854116b17f1ad557ebd
Top:    746d1fda6d9b5aadbe7969bcaa08281766ba9a13
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-06-02 11:53:48 +0100

Fix the polygon touching contour test in poly_ChkContour

This polygon must report as good:

Polygon("")
  (
[85000 50000]
[85000 90000]
[83000 90000]
[83536 63535]
[85000 59999] ##
[83535 56464]
)

The intersection forms a polygon like this:

\|  However, the right-hand edge does NOT have a node at the junction.
/|  This previously caused it to fail the self-intersection test.


This polygon must report as bad:

Polygon("")
  (
[85000 50000]
[85000 90000]
[83000 90000]
[83536 63535]
[85000 59999] ##
[89535 56464] ##
)

The intersection forms a polygon like this:

 |/  (The vertical section is a straight line with no node in the middle)
/|


---

diff --git a/src/polygon1.c b/src/polygon1.c
index 1908916..1d64929 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -3216,21 +3216,45 @@ poly_ChkContour (PLINE * a)
 	      else if (vect_dist2 (i1, a1->next->point) < EPSILON)
 		hit1 = a1->next;
 	      else
-		return TRUE;
+		hit1 = NULL;
 
 	      if (vect_dist2 (i1, a2->point) < EPSILON)
 		hit2 = a2;
 	      else if (vect_dist2 (i1, a2->next->point) < EPSILON)
 		hit2 = a2->next;
 	      else
-		return TRUE;
+		hit2 = NULL;
 
-#if 1
-	      /* now check if they are inside each other */
-	      if (inside_sector (hit1, hit2->prev->point) !=
-		  inside_sector (hit1, hit2->next->point))
-		return TRUE;
-#endif
+	      if ((hit1 == NULL) && (hit2 == NULL))
+		{
+		  /* If the intersection didn't land on an end-point of either
+		   * line, we know the lines cross and we return TRUE.
+		   */
+		  return TRUE;
+		}
+	      else if (hit1 == NULL)
+		{
+		/* An end-point of the second line touched somewhere along the
+		   length of the first line. Check where the second line leads. */
+		  if (inside_sector (hit2, a1->point) !=
+		      inside_sector (hit2, a1->next->point))
+		    return TRUE;
+		}
+	      else if (hit2 == NULL)
+		{
+		/* An end-point of the first line touched somewhere along the
+		   length of the second line. Check where the first line leads. */
+		  if (inside_sector (hit1, a2->point) !=
+		      inside_sector (hit1, a2->next->point))
+		    return TRUE;
+		}
+	      else
+		{
+		/* Both lines intersect at an end-point. Check where they lead. */
+		  if (inside_sector (hit1, hit2->prev->point) !=
+		      inside_sector (hit1, hit2->next->point))
+		    return TRUE;
+		}
 	    }
 	}
       while ((a2 = a2->next) != &a->head);
